File: scan_best_by_inmode.py
'''
Copyright © 2018 Anton Tsukanov. Contacts: tsukanov@bionet.nsc.ru
License: http://www.gnu.org/licenses/gpl.txt

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
'''

import os
import sys
import shlex
import subprocess
import argparse
import glob
import numpy as np
import pandas as pd
import math


def inmode_scan(path_to_inmode, path_java, input_data, input_model, tmp_dir,
                     fpr_for_thr=1):

    args = [path_java, '-Xmx6G', '-Xms1024m', '--add-modules', 'java.xml.bind', '-jar', path_to_inmode, 'scan',
            'i={}'.format(input_model),
            'id={}'.format(input_data),
           'f={}'.format(fpr_for_thr),
           'outdir={}'.format(tmp_dir)]
    r = subprocess.call(args)
    pass


def parse_inmode_results(inmode_bed, out):

    container = list()
    append = container.append
    df = pd.read_csv(inmode_bed, sep='\t', header=None)
    df = df.sort_values(by=[0,4])
    index = 0
    for number, line in df.iterrows():
        if line[0] != index:
            append(last_score)
        index = line[0]
        last_score = line[4]
    append(line[4])

    with open(out, 'w') as file:
        for i in container:
            file.write('{}\n'.format(math.log2(float(i))))
    file.close()
    pass
    


def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument('fasta', action='store', help='path to FASTA file')
    parser.add_argument('model', action='store', help='path to xml inmode model file')
    parser.add_argument('output', action='store', help='prth to write results')
    parser.add_argument('inmode', action='store', help='path to InMoDe')
    parser.add_argument('-j', '--java', action='store', dest='java',
                            required=False, default='java', help='path to java')
    parser.add_argument('-t', '--tmp', action='store', dest='tmp',
                            required=False, default='./tmp', help='tmp dir')
    return(parser.parse_args())


def main():

    args = parse_args()
    path_to_model = args.model
    path_to_inmode = args.inmode
    fasta_path = args.fasta
    out = args.output
    tmp_dir = args.tmp
    path_to_java = args.java

    inmode_scan(path_to_inmode, path_to_java, fasta_path, path_to_model, tmp_dir)
    inmode_bed = glob.glob(tmp_dir + '/*.BED')[0]
    parse_inmode_results(inmode_bed, out)
    os.system("rm -r {}".format(tmp_dir))

if __name__ == '__main__':
    main()

File: scan_best_by_inmode.jl
import CSV
import DataFrames
import Random
using ArgParse


function read_fasta(path)
    container = Dict("ID" => String[],
    "chr" => String[],
    "start" => Int64[],
    "end" => Int64[],
    "strand" => String[],
    "seq" => String[])
    for line in eachline(path)
        if line[1] == '>'
            container["ID"] = push!(container["ID"], split(line, "::")[1][2:end])
            second_part = split(line, "::")[2]
            container["chr"] = push!(container["chr"], split(second_part, ":")[1])
            container["start"] = push!(container["start"], parse(Int64, split(split(second_part, ":")[2][1:end-2], "-")[1]))
            container["end"] = push!(container["end"], parse(Int64, split(split(second_part, ":")[2][1:end-2], "-")[2]))
            container["strand"] = push!(container["strand"], "+")
        else
            container["seq"] = push!(container["seq"], uppercase(line))
        end
    end
    return(container)
end

function read_best_inmode(path)
    ID = 0
    score = 0.0
    container = Dict("ID" => Int64[],
    "start" => Int64[],
    "end" => Int64[],
    "strand" => String[],
    "score" => Float64[] )
    
    line_id = Int64
    line_start = Int64
    line_end = Int64
    line_strand = String
    line_score = Float64

    for line in eachline(path)
        if (parse(Int64, split(line)[1]) == ID) & (parse(Float64, split(line)[5]) > score)
            line_id = parse(Int64, split(line)[1])
            line_start = parse(Int64, split(line)[2])
            line_end = parse(Int64, split(line)[3])
            line_strand = split(line)[4]
            line_score = parse(Float64, split(line)[5])
            score = line_score
            
        elseif (parse(Int64, split(line)[1]) != ID)
            
            container["ID"] = push!(container["ID"], line_id)
            container["start"] = push!(container["start"], line_start)
            container["end"] = push!(container["end"], line_end)
            container["strand"] = push!(container["strand"], line_strand)
            container["score"] = push!(container["score"], line_score)
            
            line_id = parse(Int64, split(line)[1])
            ID = line_id
            line_start = parse(Int64, split(line)[2])
            line_end = parse(Int64, split(line)[3])
            line_strand = split(line)[4]
            line_score = parse(Float64, split(line)[5])
            score = line_score
        end
    end
    return(container)
end


function reverse_complement(site::String)
    complement = ""
    for i in site
        #println(i)
        if i == 'C'
            complement = string('G', complement)
        elseif i == 'G'
            complement = string('C', complement)
        elseif i == 'A'
            complement = string('T', complement)
        elseif i == 'T'
            complement = string('A', complement)
        end
    end
    return(complement)
end


function create_bed(fasta, inmode)
    container = Dict("ID" => String[],
    "chr" => String[],
    "start" => Int64[],
    "end" => Int64[],
    "strand" => String[],
    "site" => String[],
    "score" => Float64[])

    for (score, start, enD, strand, ID) in zip(inmode["score"], inmode["start"], inmode["end"], inmode["strand"], inmode["ID"])
        if strand == "-"
            container["site"] = push!(container["site"], reverse_complement(fasta["seq"][ID+1][start+1:enD]))
        else
            container["site"] = push!(container["site"], fasta["seq"][ID+1][start+1:enD])
        end

        container["chr"] = push!(container["chr"], fasta["chr"][ID+1])
        container["start"] = push!(container["start"], fasta["start"][ID+1] + start)
        container["end"] = push!(container["end"], fasta["start"][ID+1] + enD)   
        container["ID"] = push!(container["ID"], string("peaks_", string(ID)))
        container["strand"] = push!(container["strand"], strand)
        container["score"] = push!(container["score"], score)
    end
    
    df = DataFrames.DataFrame(chr = container["chr"],
        st = container["start"],
        en = container["end"],
        id = container["ID"],
        score = container["score"],
        strand = container["strand"],
        site = container["site"])
    
    return(df)
end


function scan_all_by_inmode(path_to_inmode, path_to_model, path_to_fasta, tmp_dir)
    #run(`/home/anton/Programs/jdk-9/bin/java -Xmx4096m -Xms1024m --add-modules java.xml.bind -jar $path_to_inmode scan
    #run(`java -jar $path_to_inmode scan
    run(`/Users/anton/Documents/Programs/jre-9.0.4.jre/Contents/Home/bin/java -Xmx4096m -Xms1024m --add-modules java.xml.bind -jar $path_to_inmode scan
    i="$path_to_model"
    id="$path_to_fasta"
    f=1.0
    outdir=$tmp_dir
    bs=true`);
end


function parse_commandline()
    s = ArgParseSettings()

    @add_arg_table s begin
        "output"
            help = "path to write results"
            required = true
        "input"
            help = "path to xml file InMoDe model"
            required = true
        "fasta"
            help = "path to fasta file with sites"
            required = true
        "inmode"
            help = "path to inmode program"
            required = true
        "--tmp", "-t"
            help = "dir for tmp files"
            required = false
            arg_type = String
            default = "./tmp"

    end
    return parse_args(s)
end


function main()
    args = parse_commandline()
    path_to_model = args["input"]
    out = args["output"]
    path_to_fasta = args["fasta"]
    tmp_dir = args["tmp"]
    path_to_inmode = args["inmode"]
    
    if !isdir(tmp_dir)
        mkdir(tmp_dir)
    end

    scan_all_by_inmode(path_to_inmode, path_to_model, path_to_fasta, tmp_dir)
    fasta = read_fasta(path_to_fasta)
    inmode_path = string(tmp_dir, "/Motif_hits_from_SequenceScan(1.0).BED")
    inmode = read_best_inmode(inmode_path)
    df = create_bed(fasta, inmode)
    rm(tmp_dir, recursive=true)
    CSV.write(out, df, delim='\t',writeheader=false)
end

main()